File: backend/src/controllers/urlController.ts
import { Request, Response, NextFunction } from 'express';
import { AuthRequest } from '../middleware/auth';
import * as urlService from '../services/urlService';
export const create = async (req: AuthRequest, res: Response, next: NextFunction) => {
  try {
    const { original_url, custom_alias, expires_at, password, is_one_time } = req.body;
    const url = await urlService.createShortUrl({
      original_url,
      custom_alias,
      expires_at,
      password,
      is_one_time,
      user_id: req.user ? req.user.id : null
    });
    res.status(201).json({
      success: true,
      data: url
    });
  } catch (error) {
    next(error);
  }
};
export const list = async (req: AuthRequest, res: Response, next: NextFunction) => {
  try {
    const { search, is_active, page, limit } = req.query;
    if (!req.user) {
      return res.status(401).json({ success: false, error: 'Unauthorized' });
    }
    let isActiveFilter: boolean | undefined = undefined;
    if (is_active !== undefined) {
      isActiveFilter = is_active === 'true';
    }
    const pageNum = parseInt(page as string, 10) || 1;
    const limitNum = parseInt(limit as string, 10) || 10;
    const data = await urlService.getUrls({
      user_id: req.user.id,
      search: search as string,
      isActive: isActiveFilter,
      page: pageNum,
      limit: limitNum
    });
    res.status(200).json({
      success: true,
      ...data
    });
  } catch (error) {
    next(error);
  }
};
export const update = async (req: AuthRequest, res: Response, next: NextFunction) => {
  try {
    const { id } = req.params as { id: string };
    const { original_url, custom_alias, expires_at, is_active, password, is_one_time } = req.body;
    if (!req.user) {
      return res.status(401).json({ success: false, error: 'Unauthorized' });
    }
    const url = await urlService.updateUrl({
      id,
      user_id: req.user.id,
      original_url,
      custom_alias,
      expires_at,
      is_active,
      password,
      is_one_time
    });
    res.status(200).json({
      success: true,
      data: url
    });
  } catch (error) {
    next(error);
  }
};
export const remove = async (req: AuthRequest, res: Response, next: NextFunction) => {
  try {
    const { id } = req.params as { id: string };
    if (!req.user) {
      return res.status(401).json({ success: false, error: 'Unauthorized' });
    }
    await urlService.deleteUrl(id, req.user.id);
    res.status(200).json({
      success: true,
      message: 'URL successfully deleted.'
    });
  } catch (error) {
    next(error);
  }
};
export const redirectUrl = async (req: any, res: Response, next: NextFunction) => {
  try {
    const { shortCode } = req.params as { shortCode: string };
    const frontendUrl = process.env.CLIENT_URL || 'http://localhost:5173';
    return res.redirect(302, `${frontendUrl}/redirect/${shortCode}`);
  } catch (error) {
    next(error);
  }
};
export const resolveShortUrl = async (req: Request, res: Response, next: NextFunction) => {
  try {
    const { shortCode } = req.params as { shortCode: string };
    const { password } = req.body;
    const ipAddress = req.ip || '';
    const userAgent = req.headers['user-agent'];
    const data = await urlService.resolveUrlAndTrackVisit(shortCode, ipAddress, userAgent, password);
    res.status(200).json({
      success: true,
      data
    });
  } catch (error) {
    next(error);
  }
};

File: backend/src/main/java/com/urlshortener/controller/UrlController.java
package com.urlshortener.controller;

import com.urlshortener.dto.request.CreateUrlRequest;
import com.urlshortener.dto.request.ResolveUrlRequest;
import com.urlshortener.dto.request.UpdateUrlRequest;
import com.urlshortener.entity.Url;
import com.urlshortener.service.UrlService;

import jakarta.servlet.http.HttpServletRequest;
import jakarta.validation.Valid;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.*;

@RestController
@RequestMapping("/api/urls")
public class UrlController {

    @Autowired
    private UrlService urlService;

    @PostMapping
    public ResponseEntity<?> createShortUrl(@Valid @RequestBody CreateUrlRequest request,
                                             HttpServletRequest httpRequest) {
        
        UUID userId = (UUID) httpRequest.getAttribute("userId");

        Url url = urlService.createShortUrl(request, userId);

        Map<String, Object> response = new LinkedHashMap<>();
        response.put("success", true);
        response.put("data", url);

        return ResponseEntity.status(HttpStatus.CREATED).body(response);
    }

    @GetMapping
    public ResponseEntity<?> getUrls(HttpServletRequest request,
                                      @RequestParam(value = "search", required = false) String search,
                                      @RequestParam(value = "is_active", required = false) Boolean isActive,
                                      @RequestParam(value = "page", defaultValue = "1") int page,
                                      @RequestParam(value = "limit", defaultValue = "10") int limit) {
        @SuppressWarnings("unchecked")
        Map<String, Object> user = (Map<String, Object>) request.getAttribute("user");
        UUID userId = (UUID) request.getAttribute("userId");

        if (user == null) {
            return ResponseEntity.status(401).body(Map.of("success", false, "error", "Unauthorized"));
        }

        Map<String, Object> result = urlService.getUrls(userId, search, isActive, page, limit);

        Map<String, Object> response = new LinkedHashMap<>();
        response.put("success", true);
        response.put("urls", result.get("urls"));
        response.put("pagination", result.get("pagination"));

        return ResponseEntity.ok(response);
    }

    @PutMapping("/{id}")
    public ResponseEntity<?> updateUrl(@PathVariable("id") UUID id,
                                        @RequestBody UpdateUrlRequest request,
                                        HttpServletRequest httpRequest) {
        @SuppressWarnings("unchecked")
        Map<String, Object> user = (Map<String, Object>) httpRequest.getAttribute("user");
        UUID userId = (UUID) httpRequest.getAttribute("userId");

        if (user == null) {
            return ResponseEntity.status(401).body(Map.of("success", false, "error", "Unauthorized"));
        }

        Url url = urlService.updateUrl(id, userId, request);

        Map<String, Object> response = new LinkedHashMap<>();
        response.put("success", true);
        response.put("data", url);

        return ResponseEntity.ok(response);
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<?> deleteUrl(@PathVariable("id") UUID id,
                                        HttpServletRequest httpRequest) {
        @SuppressWarnings("unchecked")
        Map<String, Object> user = (Map<String, Object>) httpRequest.getAttribute("user");
        UUID userId = (UUID) httpRequest.getAttribute("userId");

        if (user == null) {
            return ResponseEntity.status(401).body(Map.of("success", false, "error", "Unauthorized"));
        }

        urlService.deleteUrl(id, userId);

        Map<String, Object> response = new LinkedHashMap<>();
        response.put("success", true);
        response.put("message", "URL successfully deleted.");

        return ResponseEntity.ok(response);
    }

    @PostMapping("/resolve/{shortCode}")
    public ResponseEntity<?> resolveUrl(@PathVariable("shortCode") String shortCode,
                                         @RequestBody(required = false) ResolveUrlRequest request,
                                         HttpServletRequest httpRequest) {
        String ip = httpRequest.getRemoteAddr();
        String userAgent = httpRequest.getHeader("User-Agent");
        String password = (request != null) ? request.getPassword() : null;

        Map<String, Object> result = urlService.resolveUrlAndTrackVisit(shortCode, ip, userAgent, password);

        Map<String, Object> response = new LinkedHashMap<>();
        response.put("success", true);
        response.put("data", result);

        return ResponseEntity.ok(response);
    }
}
